design/design-doc-aggregate-resources.md

```diff
@@ -0,0 +1,965 @@
+
+* Ensures that several interdependent resources submitted to Crossplane
+  simultaneously may eventually resolve to functioning infrastructure.
+* Avoids unnecessary API churn, especially to `v1beta1` managed resource APIs
+  which must maintain backward compatibility.
+* Does not introduce the requirement that any one Crossplane controller knows
+  all extant kinds of managed resource, resource class, or resource claim.
+* Maintains the high-fidelity, composable nature of managed resources.
+
+## Proposal
+
+This document proposes Crossplane introduce two new kinds:
+`AggregateResourceClass` and `AggregateResource`:
+
+```yaml
+apiVersion: aggregation.crossplane.io/v1alpha1
+kind: AggregateResourceClass
+metadata:
+  name: cool-gke-cluster
+aggregationRule:
+  classSelector:
+    matchLabels:
+      resourceclass.crossplane.io/aggregate-to: cool-gke-cluster
+specTemplate:
+  writeConnectionSecretsToNamespace: crossplane-system
+```
+
+```yaml
+apiVersion: aggregation.crossplane.io/v1alpha1
+kind: AggregateResource
+metadata:
+  name: default-coolcluster-g3bf7
+spec:
+  aggregationRule:
+    resourceSelector:
+      matchLabels:
+        resourceclaim.crossplane.io/namespace: default
+        resourceclaim.crossplane.io/name: coolcluster
+        resourceclaim.crossplane.io/uid: eabce854-0cd7-11ea-8d71-362b9e155667
+  claimRef:
+    apiVersion: compute.crossplane.io/v1alpha1
+    kind: KubernetesCluster
+    namespace: default
+    name: coolcluster
+    uid: eabce854-0cd7-11ea-8d71-362b9e155667
+  classRef:
+    apiVersion: aggregation.crossplane.io/v1alpha1
+    kind: AggregateResourceClass
+    name: cool-gke-cluster
+    uid: ea414c74-0cd9-11ea-8d71-362b9e155667
+  writeConnectionSecretsToRef:
+    namespace: crossplane-system
+    name: eabce854-0cd7-11ea-8d71-362b9e155667
+```
+
+The purpose of these new resources, as their names imply, is to _aggregate_
+resource classes and managed resources, enabling transitive one-to-many
+resource-claim-to-resource-class and resource-claim-to-managed-resource
+relationships. Both new kinds use an _aggregation rule_, like an [RBAC
+`ClusterRole`]. The `AggregateResourceClass` aggregates resource classes (of any
+non-aggregate kind) by label, while the `AggregateResource` aggregates managed
+resources (again, of any non-aggregate kind) by label.
+
+In the above example `default-coolcluster-g3bf7` is an `AggregateResource` that
+was dynamically provisioned to satisfy a `KubernetesCluster` resource claim. An
+`AggregateResource` was called for because the `KubernetesCluster` specified an
+`AggregateResourceClass` as its `classRef`:
+
+```yaml
+apiVersion: compute.crossplane.io/v1beta1
+kind: KubernetesCluster
+metadata:
+  namespace: default
+  name: coolcluster
+spec:
+  resourceRef:
+    apiVersion: aggregation.crossplane.io/v1alpha1
+    kind: AggregateResource
+    name: default-coolcluster-g3bf7
+    uid: 2d45b85e-0cdd-11ea-8d71-362b9e155667
+  classRef:
```

Comment:
It would. I imagine only an `AggregateResourceClass` that aggregated a compatible resource class would be eligible for matching; i.e. an `AggregateResourceClass` would need to aggregate a `GKECluster` (or EKS etc) to be eligible to be selected for a `KubernetesCluster` claim.